### ISSUE TITLE
avoid duplicate nuget packages

### ISSUE THREAD

**Issue by cwuethrich** [nuke-build/nuke#1344]
### Usage Information

7.0.6 / 8.0.101  / net8.0 / Windows

### Description

When you install a .NET tool as a [local tool](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool), two `.nupkg` files appear in the package folder. This might be a bug in dotnet, but it would be challenging to get this fixed.

The package resolver of NUKE expects to find only one package of the same version. We should change this behavior to allow multiple files of the same version and ignore the duplicates.

For example, for dotnetsay, you can find `dotnetsay.2.1.7.nupkg` and `dotnetsay.nupkg` under the folder `%USERPROFILE%\.nuget\packages\dotnetsay\2.1.7`.

### Reproduction Steps

Make sure this folder `%USERPROFILE%\.nuget\packages\swashbuckle.aspnetcore.cli\6.5.0` doesn't exist.

`dotnet new tool-manifest`
`dotnet tool install --version 6.5.0 Swashbuckle.AspNetCore.Cli`

The package was downloaded to the folder `%USERPROFILE%\.nuget\packages\swashbuckle.aspnetcore.cli\6.5.0`, and the two files `swashbuckle.aspnetcore.cli.6.5.0.nupkg` and `Swashbuckle.AspNetCore.Cli.nupkg` appear in this folder.

### Expected Behavior

Nuke should ignore the file `Swashbuckle.AspNetCore.Cli.nupkg`, as it is a copy of `swashbuckle.aspnetcore.cli.6.5.0.nupkg`.

### Actual Behavior

Nuke throws an exception "Package 'Swashbuckle.AspNetCore.Cli' is referenced with multiple versions. Use NuGetPackageResolver and SetToolPath." with inner exception "Sequence contains more than one matching element".

### Regression?

I don't think this case was handled before.

### Known Workarounds

I will attach a pull request, as there is no possible workaround.

### Could you help with a pull-request?

Yes

**Comment by mlaily** [comment]
I believe I'm hitting the problem described here.

Actually, just calling `NuGetPackageResolver.GetLocalInstalledPackages(NuGetToolPathResolver.NuGetAssetsConfigFile).ToList()` is enough to trigger an exception because the `.SingleOrDefault()` has multiple matching items in my current .net 8 solution:
![image](https://github.com/nuke-build/nuke/assets/1535823/2932dbee-be7f-4bfd-b547-ee7371947a4e)

**Comment by willsmith9182** [comment]
i've been wondering if there's any scope to lean on [local dotnet tools](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-local-tool) for cli tools rather than including them as a package reference?

we currently manually call `dotnet tool restore` as part of our build and then use it as a cli tool directly. We plan to start making our own tool.json to point at the cli tool over the package tool and provide a strongly typed api over using the fluent api for invoking tools. 

the there something gained when using `<PackageReference>` that i'm missing?

**Comment by matkoch** [comment]
It's not clear to me why `tool-manifest` is used. The resolver does not consider it. Therefore I'm expecting it's duplicated in the `csproj` file. I'm not saying this duplicates the file, but that tool manifests are unnecessary from my point of view. @ChristophLindemann also mentioned in the other issue that the packages are duplicated in `.dotnet/tools/.store` – that also is irrelevant.

Bottom line, I'd like to hear a good argument why tool manifests are being used, and not just `PackageDownload`. I couldn't reproduce the issue by just using `PackageDownload`.

**Comment by mlaily** [comment]
I'm not sure I understand everything you said, but in my case I was just trying to use [paket](https://fsprojects.github.io/Paket/) through `PaketTasks.PacketPack`, and it failed with the error mentioned in this issue.

As my solution is pretty big I'm honestly not sure what is causing this behaviour (duplicate packages), and it might even be unrelated to paket itself, but the result is that I can't use paket from my nuke script without resorting to calling it entirely manually. (I'm currently calling it through `DotNetTasks.DotNet("paket pack ...")` instead as a workaround)

Installing the `paket` dotnet tool is the standard and normal way to use it in a project, and I was hoping I could simply call the already installed tool through the nuke helpers in my build script.

It would be weird to reference paket a second time in my build.csproj.

**Comment by matkoch** [comment]
> It would be weird to reference paket a second time in my build.csproj.

that's the fundamental difference in our perspectives. i think the build project is the best place to define versions of tools that are required. otherwise it's not self-contained and may lead to change in behavior on different machines.

**Comment by ChristophLindemann** [comment]
> It's not clear to me why `tool-manifest` is used. The resolver does not consider it. Therefore I'm expecting it's duplicated in the `csproj` file. I'm not saying this duplicates the file, but that tool manifests are unnecessary from my point of view. @ChristophLindemann also mentioned in the other issue that the packages are duplicated in `.dotnet/tools/.store` – that also is irrelevant.
> 
> Bottom line, I'd like to hear a good argument why tool manifests are being used, and not just `PackageDownload`. I couldn't reproduce the issue by just using `PackageDownload`.

@matkoch I am on vacation right now, so I can not tell you specifically why it did not work. I can look into it in about 14 days. 
The problem occurred in projects A if I had other unrelated projects B that where restored with the new .NET 8.0 SDK version (mentioned in #1386) This other project might restore the "double" nupkg file, and then the NUKE resolver would fail for project A. (if I remember correctly) It worked fine on most machines, and then suddenly the builds would fail on some build machines.

**Comment by cwuethrich** [comment]
> Bottom line, I'd like to hear a good argument why tool manifests are being used, and not just `PackageDownload`. I couldn't reproduce the issue by just using `PackageDownload`.

NUKE itself doesn't cause a duplicate of the package. It is, in fact, when you install a dotnet tool as a local tool that you end up with two identical files in the package folder. In my case, we use Swashbuckle in older projects where we didn't use NUKE. We don't install it as a global tool because we want to be able to reproduce the exact same behavior (use the same version).

For new projects, we started using NUKE. Because of the incorrect behavior of `dotnet restore`, we had to include our own versions of the `PackageResolver` to avoid encountering this exception.

The reason for NUKE to change this behavior of the `PackageResolver` is to achieve good compatibility with other build processes.
